metagraph/src/annotation/annotation_converters.cpp

```diff
@@ -978,5 +980,111 @@ void convert_to_row_annotator(const ColumnCompressed<std::string> &source,
+    Vector<uint64_t> node_diffs;
+    Vector<std::pair<uint64_t, uint32_t>> pos(nnodes, { 0, 0 });
+    sdsl::bit_vector terminal(nnodes, 0);
+
+    std::atomic<uint64_t> terminal_count = 0;
+    std::atomic<uint64_t> forced_terminal_count = 0;
+    std::atomic<uint64_t> depth_terminal_count = 0;
+    std::atomic<uint64_t> boundary_size = 0;
+    std::atomic<uint64_t> visited_nodes = 0;
+    graph.call_sequences(
+        [&](const std::string &, const std::vector<uint64_t> &path) {
+            assert(!path.empty());
+            std::vector<uint64_t> anno_ids(path.size());
+            for (uint32_t i = 0; i < path.size(); ++i) {
+                assert(path[i] <= nnodes);
+                anno_ids[i] = graph::AnnotatedSequenceGraph::graph_to_anno_index(path[i]);
+            }
+            std::vector<Vector<uint64_t>> rows = annotation.get_matrix().get_rows(anno_ids);
+            visited_nodes += path.size();
+            std::sort(rows[0].begin(), rows[0].end());
+            Vector<uint64_t> diffs;
+            for (uint32_t i = 0; i < rows.size() - 1; ++i) {
+                std::sort(rows[i + 1].begin(), rows[i + 1].end());
+
+                // if we reached the max path depth, force a terminal node
+                if (i % max_depth == 0) {
+                    pos[anno_ids[i]] = { diffs.size(), rows[i].size() };
+                    diffs.insert(diffs.end(), rows[i].begin(), rows[i].end());
+                    terminal[anno_ids[i]] = 1;
+                    depth_terminal_count.fetch_add(1, std::memory_order_relaxed);
+                    continue;
+                }
+
+                uint64_t start_idx = diffs.size();
+                std::set_symmetric_difference(rows[i].begin(), rows[i].end(),
+                                              rows[i + 1].begin(), rows[i + 1].end(),
+                                              std::back_inserter(diffs));
+
+                // if we don't gain anything by diffing, make the node terminal
+                if (diffs.size() - start_idx >= rows[i].size()) {
+                    diffs.resize(start_idx);
+                    diffs.insert(diffs.end(), rows[i].begin(), rows[i].end());
+                    terminal[anno_ids[i]] = 1;
+                    forced_terminal_count.fetch_add(1, std::memory_order_relaxed);
+                }
+                pos[anno_ids[i]] = { start_idx, diffs.size() - start_idx };
+            }
+            pos[anno_ids.back()] = { diffs.size(), rows.back().size() };
+            diffs.insert(diffs.end(), rows.back().begin(), rows.back().end());
+            size_t offset;
+#pragma omp critical
+            {
+                offset = node_diffs.size();
+                node_diffs.insert(node_diffs.end(), diffs.begin(), diffs.end());
+            }
+            for (uint32_t i = 0; i < rows.size(); ++i) {
+                pos[anno_ids[i]].first += offset;
+            }
+
+            terminal_count.fetch_add(1, std::memory_order_relaxed);
```

Comment:
this is a really negligible computation, I would just use atomic variables to make it easier to read

Reply:
I suggested this change because there are several atomic variables in this section which are being incremented frequently, so using `operator++` causes their memory order to be `std::memory_order_seq_cst`. In some contexts, this can lead to performance loss because we are unnecessarily forcing all of these operations to be performed in order.

In this context, however, the graph traversal, `get_rows`, and `sort` calls dominate the runtime, so most likely this change won't lead to a performance boost. Using `fetch_add` just gives a hint to the reader that we there are no dependences between these increments.